deepset_cloud_sdk/workflows/pipeline_client/pipeline_service.py

```diff
@@ -0,0 +1,257 @@
+        :param config: Configuration for publishing
+        """
+        pipeline_yaml = self._create_config_yaml(pipeline, config)
+        response = await self.api.post(
+            workspace_name=DEFAULT_WORKSPACE_NAME,
+            endpoint="indexes",
+            json={"name": config.name, "config_yaml": pipeline_yaml},
+        )
+        response.raise_for_status()
+        if response.status_code == 201:
+            logger.info(f"Index {config.name} successfully created")
+
+    async def publish(self, pipeline: PipelineProtocol, config: PublishConfig) -> None:
+        """Publish a pipeline or indexto deepset AI platform.
+
+        :param pipeline: The pipeline or index to publish. Must implement the PipelineProtocol
+        :param config: Configuration for publishing
+
+        :raises TypeError: If the pipeline object doesn't implement the required protocol
+        :raises ValueError: If no workspace is configured
+        """
+        logger.info(f"Starting publish process for {config.pipeline_type.value}: {config.name}")
+
+        if not isinstance(pipeline, PipelineProtocol):
+            raise TypeError(PipelineServiceDocs.INVALID_PIPELINE_TYPE_ERROR)
+
+        if not DEFAULT_WORKSPACE_NAME:
+            raise ValueError(PipelineServiceDocs.WORKSPACE_REQUIRED_ERROR)
+
+        logger.info(f"Publishing {config.pipeline_type.value} to workspace {DEFAULT_WORKSPACE_NAME}")
+
+        if config.pipeline_type == PipelineType.INDEX:
+            await self._publish_index(pipeline, config)
+        else:
+            await self._publish_pipeline(pipeline, config)
+
+
+def enable_publish_to_deepset() -> None:
+    """Add the publish method to the Haystack Pipeline and AsyncPipeline classes if not already added.
+
+    Usage:
+
+    Run `deepset-cloud login` and follow the instructions to authenticate to deepset AI platform.
+
+    ```python
+    from haystack import Pipeline
+    from deepset_cloud_sdk.workflows import enable_publish_to_deepset
+    enable_publish_to_deepset()
+
+    # define your Haystack Pipeline
+    pipeline = Pipeline()
+
+    # add components to the pipeline and connect them...
+
+    # define your publish config
+    publish_config = PublishConfig(
+        name="my-pipeline",
+        pipeline_type=PipelineType.PIPELINE,
+        inputs=PipelineInputs(
+            query=["chat_summary_prompt_builder.question"]),
+            filters=["bm25_retriever.filters", "embedding_retriever.filters"]
+        ),
+        outputs=PipelineOutputs(
+            documents="ranker.documents",
+            answers="ranker.answers"
+        )
+    )
+    asyncio.run(pipeline.publish(pipeline=pipeline, config=publish_config))
+    ```
+    """
+    try:
+        from haystack import AsyncPipeline as HaystackAsyncPipeline
+        from haystack import Pipeline as HaystackPipeline
+
+        async def publish_to_deepset(self: PipelineProtocol, config: PublishConfig) -> None:
```

Comment:
not sure if I'd use `async` here. Might actually rather be in the way for developers because they then need a loop and everything becomes complicated

Reply:
I implemented an async and a sync client, where the sync one is just calling "asyncio.run" on the async implementation to support both. For pipelines this makes sense, but for files it was useful.  Not 100% sure if we want to stick to this

Reply:
we could also offer a sync `publish` and an async `apublish` or `publish_async`?